src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,227 @@
+
+# if __name__ == "__main__":
+#     main()
+
+
+def connect_to_api(url, method="GET", headers=None, params=None, data=None):
+    """
+    Connect to an API using the requests library.
+
+    :param url: URL of the API endpoint.
+    :param method: HTTP method (e.g., 'GET', 'POST', 'PUT', 'DELETE', etc.). Default is 'GET'.
+    :param headers: Dictionary of HTTP headers to send with the request.
+    :param params: Dictionary of URL parameters to append to the URL.
+    :param data: Dictionary of data to send in the body of the request (for POST, PUT, DELETE).
+    :return: Response object from requests.
+    """
+    try:
+        response = requests.request(
+            method, url, headers=headers, params=params, data=data
+        )
+        if response.status_code == 429:
+            # Extract the Retry-After header and sleep for that duration
+            retry_after = int(
+                response.headers.get("Retry-After", 10)
+            )  # Default to 60 seconds if header is missing
+            print(f"Rate limit hit. Retrying after {retry_after} seconds.")
+            time.sleep(retry_after)
+        else:
+            return response
+    except requests.exceptions.HTTPError as errh:
+        print(f"Http Error: {errh}")
+    except requests.exceptions.ConnectionError as errc:
+        print(f"Error Connecting: {errc}")
+    except requests.exceptions.Timeout as errt:
+        print(f"Timeout Error: {errt}")
+    except requests.exceptions.RequestException as err:
+        print(f"Error: {err}")
+
+
+def get_data_meteo_api(city, VARIABLES):
+    """
+    Connect to an API using the requests library.
+
+    :param city: URL of the API endpoint.
+    """
+
+    # Define Climate API URL
+    API_URL = "https://climate-api.open-meteo.com/v1/climate?"
+
+    # Define coordinate of available cities
+    COORDINATES = {
+        "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+        "London": {"latitude": 51.507351, "longitude": -0.127758},
+        "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+    }
+
+    # Define variables to retrieve from API
+    # VARIABLES = ["temperature_2m_mean", "precipitation_sum", "soil_moisture_0_to_10cm_mean"]
+
+    # Define API parameters
+    params = {
+        "latitude": COORDINATES[city]["latitude"],
+        "longitude": COORDINATES[city]["longitude"],
+        "start_date": "1950-01-01",
+        "end_date": "2050-12-31",
+        "models": [
+            "CMCC_CM2_VHR4",
+            "FGOALS_f3_H",
+            "HiRAM_SIT_HR",
+            "MRI_AGCM3_2_S",
+            "EC_Earth3P_HR",
+            "MPI_ESM1_2_XR",
+            "NICAM16_8S",
+        ],
+        "daily": VARIABLES,
+    }
+
+    # Call API
+    response = connect_to_api(API_URL, params=params)
+    data = response.json()
+
+    # Create a DataFrame with the retrieved data
+    columns_df = list(data["daily_units"].keys())
+
+    dicts = {}
+    for column in columns_df:
```

Comment:
you can create a data frame directly from a `dict` (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.from_dict.html) no need to do this